### ISSUE TITLE
Allow users to arbitrarily modify the scope of the global search from the front-end UI

### ISSUE THREAD

**Issue by cpmills1975** [nautobot/nautobot#39]
### Environment
* Python version: 3.9
* Nautobot version: v1.0.0b

### Proposed Functionality
Relating to #38, users should be able to choose what models to include in the scope of their searches. A users' preference for the models to be searched should be saved for future searches.

My proposal is to replace the type selection widget in the search page with a styled multiple selection widget that, to conserve space on the home page, is only displayed when the user clicks an 'advanced search' button. The additional selection widget does not fit well within the navigation bar search, so this would be refactored to search only the previously selected set of models - a 'quick search' if you like.

### Use Case
Some may choose not to use all models within the Nautobot tool which are included in the global search scope, but make more use of models which are not included. There is little point using processor cycles searching empty tables, when populated tables containing useful data cannot be searched using the global search functionality. Users should be free to choose a list of multiple models to be searched depending on their preferences and organisational use of the tool.

*Note:* An unaccepted FR exists for this in the forerunner project. I have proof of concept code for this functionality developed already which could trivially be refactored and ported to Nautobot. I'd be happy to submit a PR for Nautobot if this feature is accepted and assigned to me.

### Database Changes
None

### External Dependencies
None

**Comment by jathanism** [comment]
At first glance this seems like it would be simple enough, but something to consider is that the mapping of what objects and fields get included in global search is hard-coded in [`nautobot.core.constants.SEARCH_TYPES`](https://github.com/nautobot/nautobot/blob/develop/nautobot/core/constants.py#L60-L259). And then adding the type selection to the dropdown selection is hard-coded in [`nautobot.core.forms`](https://github.com/nautobot/nautobot/blob/develop/nautobot/core/forms.py#L5-L45).

Inspecting those you can see there are a large number of objects and database field relationships to other objects that needs to be considered for efficient and intuitive search. Moving this to a user-definable interface would require that the search criteria be stored in a way that is more easily defined, and that is not as straightforward as it sounds.

To do it right, we'd need to vet every single object that we select to be eligible for search, and then resign the interface. The entire implementation of global search would need to be rethought or possible replaced altogether with something more specialized for search, with a search index, so as to not result in a horrible and laggy user experience.

I'm personally not opposed to this as a longer term solution, but I think it would be good to track some use-cases outside of inventory items to keep the conversation going.

**Comment by cpmills1975** [comment]
So to be clear, I'm not proposing changing how the search logic itself works, just a) including Inventory Items within the choosable types by including that within the constants.SEARCH_TYPES data structure (because that's the problem I'm ultimately trying to solve for my users*) and b) allowing the user to select more than one choosable type for the action of searching at the same time. e.g. inventory items and devices (because they both have asset tags and at the time of searching I may not know whether a specific asset tag has been assigned to an inventory item or a device).

My proposed implementation of this (PR ready for NetBox, could easily be repurposed for Nautobot), simply replaces the single object type selection with a multi-select widget on the home and /search pages and then saves the users' choice in their preferences. This was as an attempt to address a concern that adding inventory items would slow the application down too much. Personally I don't have any need to search circuits or providers as I don't use this part of the application, so I'd rather take them out of the search completely and replace them with Inventory Items. Equally searching for something you know to be an asset tag doesn't make much sense when applied to IP addresses! Clearly my use-case won't be shared by others, hence my desire to allow users to select multiple object types to include within the search at the point of use. 

There could perhaps be scope to examine in more depth the whole of the search logic, but my objective at this stage is to simply include Inventory Items within the scope of the global search.

They say a picture speaks a thousand words, please take a look at the proposed screenshots below:

![Screenshot 2021-02-26 at 21 20 32](https://user-images.githubusercontent.com/1878544/109356456-a8e63580-7878-11eb-8653-3ce89d37a436.png)

The proposed 'home page' search widget - with added drop down for 'advanced search'

![Screenshot 2021-02-26 at 21 20 46](https://user-images.githubusercontent.com/1878544/109356538-c5826d80-7878-11eb-8d44-488190c5b028.png)

The expanded 'home page' search widget showing the user has selected Devices and Inventory Items to be included in the scope of the search.

![Screenshot 2021-02-26 at 21 21 03](https://user-images.githubusercontent.com/1878544/109356602-da5f0100-7878-11eb-86f4-4c422f4d3526.png)

The navigation bar search box always uses the last selected object types (as the space doesn't lend itself to including an object type field), so is renamed "Quick Search" to indicate it's limitation

![Screenshot 2021-02-26 at 21 21 37](https://user-images.githubusercontent.com/1878544/109356740-08dcdc00-7879-11eb-8be7-d6cd9102a29a.png)

The refactored search results page showing only the object types searched for and where users can change the search criteria for the next and future searches.

* My users nearly started adding line cards as devices as their perception was that they couldn't search for them by asset tag. Of course they could by going to the inventory items tab, but because they are not included in searches performed through the global search form, they had made the incorrect assumption that they were unable to search for them.

**Comment by jathanism** [comment]
Yeah this is making sense to me and looks well thought out. If it's not too much trouble for you to submit a compatible PR for Nautobot, I'd like to ask you to go ahead and do it. The screen shots you showed seem promising and it's worth having the conversation IMO.

**Comment by jathanism** [comment]
@cpmills1975 I've been thinking about this idea and the more I think about it, the more I really would like to implement it. My only hangup is to go all the way and make the search functionality use real searching with a proper search index, such as [Haystack](http://haystacksearch.org).

Looking forward to what you've done so we can see how it shakes out.

**Comment by jedelman8** [comment]
@cpmills1975 any plans to pursue this further on your side?

**Comment by lampwins** [comment]
Also adding in here that I don't think we have yet captured the need for plugins to register models to the global search. Obviously beyond the fundamental scope of this issue, but I don't see us doing it without this one?

**Comment by gsnider2195** [comment]
We do support selecting a single type for the global search on the homepage, this would just be changing that to multi-select.

**Comment by itdependsnetworks** [comment]
We allow searchable_models for [apps](https://docs.nautobot.com/projects/core/en/stable/development/apps/api/models/global-search) we should allow to override such settings in core.

**Comment by itdependsnetworks** [comment]
We should review the individual use cases described above when implementing. So far there is

* Add model to be included
* Add field within already included model
* Add Custom Field of model to be included.
